domain/modelconfig/state/controllerstate.go

```diff
@@ -0,0 +1,127 @@
+	return &ControllerState{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// SetModelSecretBackend sets the secret backend for the given model.
+func (s *ControllerState) SetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID, backendName string) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	q := `
+SELECT b.uuid AS &SecretBackendInfo.uuid
+FROM   secret_backend b
+WHERE  b.name =
+    CASE $SecretBackendInfo.name
+    WHEN 'auto' THEN
+        CASE (
+            SELECT mt.type FROM model_type mt
+            JOIN   model m on mt.id = m.model_type_id
+            WHERE  m.uuid = $SecretBackendInfo.model_uuid
+        )
+        WHEN 'iaas' THEN 'internal'
+        WHEN 'caas' THEN 'kubernetes'
+        END
+    ELSE
+        $SecretBackendInfo.name
+    END
+`
+	backendInfo := SecretBackendInfo{Name: backendName, ModelUUID: modelUUID.String()}
+	stmt, err := s.Prepare(q, backendInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	modelBackendUpsert := `
+INSERT INTO model_secret_backend(model_uuid, secret_backend_uuid)
+VALUES ($SecretBackendInfo.model_uuid, $SecretBackendInfo.uuid)
+ON CONFLICT(model_uuid) DO UPDATE SET
+    secret_backend_uuid = EXCLUDED.secret_backend_uuid`
+	modelBackendUpsertStmt, err := s.Prepare(modelBackendUpsert, backendInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err = tx.Query(ctx, stmt, backendInfo).Get(&backendInfo)
+		if errors.Is(err, sql.ErrNoRows) {
+			return fmt.Errorf("%w: %q", backenderrors.NotFound, backendName)
+		}
+		if err != nil {
+			return fmt.Errorf("querying secret backends: %w", err)
+		}
+		err = tx.Query(ctx, modelBackendUpsertStmt, backendInfo).Run()
+		if database.IsErrConstraintForeignKey(err) {
+			return fmt.Errorf("%w: model %q", modelerrors.NotFound, modelUUID)
+		}
+		if err != nil {
+			return fmt.Errorf("setting secret backend %q for model %q: %w", backendName, modelUUID, err)
+		}
+		return nil
+	})
+	return domain.CoerceError(err)
+}
+
+// GetModelSecretBackendName returns the secret backend name
+// for a given model uuid.
+func (s *ControllerState) GetModelSecretBackendName(ctx context.Context, modelUUID coremodel.UUID) (string, error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	backendInfo := SecretBackendInfo{ModelUUID: modelUUID.String()}
+
+	stmt, err := s.Prepare(`
+SELECT sb.name AS &SecretBackendInfo.name
+FROM   model_secret_backend msb
+JOIN   secret_backend sb ON sb.uuid = msb.secret_backend_uuid
+WHERE  model_uuid = $SecretBackendInfo.model_uuid`, backendInfo)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, backendInfo).Get(&backendInfo)
+		if errors.Is(err, sql.ErrNoRows) {
+			backendInfo.Name = provider.Auto
```

Comment:
When we create the model, the backend should be `auto` by default, we should have just set the default value to either the internal backend backend uuid or the kubernetes backend uuid. When the user sets back the model secret backend from a vault backend to auto, Juju should also reset back the config to either internal or kubernetes.
So it should never be a not found here.

Reply:
Yes, this should never happen. The code here is being defensive. I could return a not found error instead.